junit-platform-launcher/src/main/java/org/junit/platform/launcher/core/DefaultLauncher.java

```diff
@@ -69,6 +71,27 @@
 		return testEngines;
 	}
 
+	/**
+	 * @return {@code true} if the graph does <em>not</em> contain a cycle; else {@code false}.
+	 */
+	static boolean isAcyclic(TestDescriptor root) {
+		Set<UniqueId> visited = new HashSet<>();
+		visited.add(root.getUniqueId());
+		Deque<TestDescriptor> stack = new ArrayDeque<>();
+		stack.push(root);
+		while (!stack.isEmpty()) {
+			for (TestDescriptor child : stack.pop().getChildren()) {
+				if (!visited.add(child.getUniqueId())) {
+					return false; // id already known: cycle detected!
+				}
+				if (child.isContainer()) {
+					stack.add(child);
```

Comment:
Using `pop()` and `add()` will reverse the iteration order. Shouldn't hurt, but might make debugging harder.

Reply:
@marcphillip Your saying that has prompted me to think more, and I think you're almost right; AFAICT, it will actually cause the tree to be traversed in _depth-first order_ rather than reversing the iteration order. As you said, it wouldn't hurt (except for debugging), but I think swapping the stack out for a `Queue<> queue = new ArrayDeque<>()` should fix things.

Reply:
You‘re right, it‘s indeed depth-first. However since we‘re visiting all descriptors anyway it doesn’t really matter. 😉